keps/sig-api-machinery/2887-openapi-enum-types/README.md

```diff
@@ -0,0 +1,787 @@
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+As a sub-KEP of [KEP-2896 OpenAPI v3](https://github.com/kubernetes/enhancements/issues/2896),
+this KEP defines Enum Types support for OpenAPI. Currently, types in our API have fields
+that are actually enums but represented as plain string. This KEP proposes a `+enum`
+annotation to type aliases that represent enums. The OpenAPI generator will have
+the capability to recognize the annotation and auto-detect possible values for an enum. 
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, types that are actually enums are represented as plain strings. For example,
+here is the definition of `Protocol`, which must be one of `TCP`, `UDP`, and `SCTP`.
+```go
+package core
+// ...
+
+// Protocol defines network protocols supported for things like container ports.
+type Protocol string
+
+const (
+	// ProtocolTCP is the TCP protocol.
+	ProtocolTCP Protocol = "TCP"
+	// ProtocolUDP is the UDP protocol.
+	ProtocolUDP Protocol = "UDP"
+	// ProtocolSCTP is the SCTP protocol.
+	ProtocolSCTP Protocol = "SCTP"
+)
+
+// ContainerPort represents a network port in a single container
+type ContainerPort struct {
+  // ...
+
+  // Required: Supports "TCP", "UDP" and "SCTP"
+  // +optional
+  Protocol Protocol
+  
+  // ...
+}
+```
+
+In order to convey that the `Protocol` type is an enum, every field of the type 
+have to comment its values like `ContainerPort` does. The same goes for validation - 
+protocol being one of "TCP", "UDP", and "SCTP" is duplicated among the code base.
+
+Tools and UIs can also benefit from the additional piece of information. For example,
+Protocol can be shown as a dropdown of possible values instead of a textbox.
+The user no longer needs to see an error of invalid value after submits the request.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Add an annotation that indicates a field as an enum.
+- Make the Kubernetes OpenAPI generator recognize the annotation.
+- Auto-detect possible values of an enum types.
```

Comment:
Call out that we're doing the migration of the Kubernetes native types? Maybe "Detect and automatically add the enum annotation to all existing enums in the Kubernetes native types"?  If there are any design details of interest around this, they could be added to this KEP as well.

Reply:
Added a section on the auto-detection of builtin types.